### DISCUSSION TITLE
Prospectus

### DISCUSSION THREAD

**Discussion by PaulWAyers** [theochem/matrix-permanent#2] [Ideas]
Proposed Workflow:

0. Functions for hard-coded (2xm), combinatoric, Glynn, and Ryser algorithm. (Pure C with external libraries.) Note that a permanent of a zero-by-zero matrix should be defined to be 1.0.
1. Wrapping these; so you can switch between them based on matrix dimension. Extra code to figure out the optimal algorithm. Python interface (Pybind11?). 
2. Writing a coupled cluster overlap (of some flavor). This would be basically translating Python code. 

From @msricher : 
There's a low-level C part, i.e. the inner loop math, some C++ abstractions for writing the overall algorithm for a general type, and handling of object types if desired.

**Comment by PaulWAyers** [comment]
Immediate tasks:
1. Write "hard-coded" 2xn and 3xn permanents. 
2. Write pseudocode for the combinatoric algorithm without Gray Code Order; upgrade to Gray Code Order; see whether "internal update" based on Gray Code order is useful (you seem to need to treat a case where a zero matrix element is swapped out special, and that conditional inside a loop might be too expensive). 
3. Write combinatoric code.
4. Write Glynn algorithm and Glynn code.
5. Write Ryser algorithm and Ryser code.
6.

**Comment by msricher** [comment]
https://www.jjj.de/fxt/fxtbook.pdf

See page 293-294 for the permutation algorithm. The location of the code in the library is also given.

Here's the library:
https://www.jjj.de/fxt/fxtpage.html

The Glynn algorithm can be directly translated from the Python code.

For the Ryser algorithm, see section 8.2.2, page 206, for the algorithm for generate all subsets. Location of the code in the library is also given.

@cassmasschelein

**Reply by msricher** [reply]
https://github.com/theochem/matrix-permanent/tree/example

I added an example of the code structure where the fxt library is included in the source file.

See src/ and the Makefile.
